packages/flutter_tools/test/general.shard/plugins_test.dart

```diff
@@ -1148,6 +1156,105 @@ flutter:
       });
 
+      testUsingContext('generateMainDartWithPluginRegistrant', () async {
+        when(flutterProject.isModule).thenReturn(false);
+
+        final List<Directory> directories = <Directory>[];
+        final Directory fakePubCache = fs.systemTempDirectory.childDirectory('cache');
+        final File packagesFile = flutterProject.directory
+            .childFile('.packages')
+            ..createSync(recursive: true);
+
+        final Map<String, String> plugins = <String, String>{};
+        plugins['url_launcher_macos'] = '''
+  flutter:
+    plugin:
+      implements: url_launcher
+      platforms:
+        macos:
+          dartPluginClass: MacOSPlugin
+''';
+        plugins['url_launcher_linux'] = '''
+  flutter:
+    plugin:
+      implements: url_launcher
+      platforms:
+        linux:
+          dartPluginClass: LinuxPlugin
+''';
+        plugins['url_launcher_windows'] = '''
+  flutter:
+    plugin:
+      implements: url_launcher
+      platforms:
+        windows:
+          dartPluginClass: WindowsPlugin
+''';
+
+        for (final MapEntry<String, String> entry in plugins.entries) {
+          final String name = fs.path.basename(entry.key);
+          final Directory pluginDirectory = fakePubCache.childDirectory(name);
+          packagesFile.writeAsStringSync(
+              '$name:file://${pluginDirectory.childFile('lib').uri}\n',
+              mode: FileMode.writeOnlyAppend);
+          pluginDirectory.childFile('pubspec.yaml')
+              ..createSync(recursive: true)
+              ..writeAsStringSync(entry.value);
+          directories.add(pluginDirectory);
+        }
+
+        when(flutterManifest.dependencies).thenReturn(<String>{...plugins.keys});
+
+        final Directory libDir = flutterProject.directory.childDirectory('lib');
+        libDir.createSync(recursive: true);
+
+        final File mainFile = libDir.childFile('main.dart');
+        mainFile.writeAsStringSync('''
+// @dart = 2.8
+void main() {
+}
+''');
+        final File flutterBuild = flutterProject.directory.childFile('generated_main.dart');
+        final bool didGenerate = await generateMainDartWithPluginRegistrant(
+          flutterProject,
+          'main',
+          flutterBuild,
+          mainFile,
+        );
+        expect(didGenerate, isTrue);
+        expect(flutterBuild.readAsStringSync(),
+            '//\n'
+            '// Generated file. Do not edit.\n'
+            '//\n'
+            '\n'
+            '// @dart = 2.8\n'
+            '\n'
+            'import \'main\' as entrypoint;\n'
+            'import \'dart:io\'; // ignore: dart_io_import.\n'
+            'import \'package:url_launcher_linux${fs.path.separator}url_launcher_linux.dart\';\n'
+            'import \'package:url_launcher_macos${fs.path.separator}url_launcher_macos.dart\';\n'
+            'import \'package:url_launcher_windows${fs.path.separator}url_launcher_windows.dart\';\n'
+            '\n'
+            '@pragma(\'vm:entry-point\')\n'
```

Comment:
I don't understand why you need an additional entrypoint when you're wrapping main anyway. Couldn't `main()` below just be:
```
void main() {
  _registerPlugins();
  entrypoint.main();
}
```

Reply:
The embedding API in the engine allows to change the main entrypoint function. As a result, there's no way to know which entrypoint is invoked at runtime. It turns out that `main` is the default.